File: 0076-minimum-window-substring/0076-minimum-window-substring.java
import java.util.HashMap;

class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";

        HashMap<Character, Integer> hash = new HashMap<>();
        HashMap<Character, Integer> need = new HashMap<>();

        for (char c : t.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }

        int required = t.length(); 
        int formed = 0; 
        int left = 0;
        int n = s.length();
        int mini = Integer.MAX_VALUE;
        int start = 0;

        for (int right = 0; right < n; right++) {
            char c = s.charAt(right);
            
            hash.put(c, hash.getOrDefault(c, 0) + 1);

            if (need.containsKey(c) && hash.get(c) <= need.get(c)) {
                formed++;
            }

            while (formed == required) {
                if (right - left + 1 < mini) {
                    mini = right - left + 1;
                    start = left;
                }

                char lefty = s.charAt(left);
                
                if (need.containsKey(lefty) && hash.get(lefty).equals(need.get(lefty))) {
                    formed--;
                }
                
                hash.put(lefty, hash.get(lefty) - 1);
                left++;
            }
        }

        return mini == Integer.MAX_VALUE
                ? ""
                : s.substring(start, start + mini);
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.go
func minWindow(s string, t string) string {

	if len(s) < len(t) {
		return ""
	}

	// Required frequencies
	need := make(map[byte]int)
	for i := 0; i < len(t); i++ {
		need[t[i]]++
	}

	// Current window frequencies
	window := make(map[byte]int)

	left := 0
	formed := 0                 // total matched characters
	required := len(t)          // total required characters

	minLen := int(1e9)
	start := 0

	for right := 0; right < len(s); right++ {

		c := s[right]

		// ALWAYS add to window
		window[c]++

		// Increase formed ONLY if this char was still needed
		if need[c] > 0 && window[c] <= need[c] {
			formed++
		}

		// Shrink only when fully formed
		for formed == required {

			if right-left+1 < minLen {
				minLen = right - left + 1
				start = left
			}

			leftChar := s[left]
			window[leftChar]--

			// If removal breaks requirement, reduce formed
			if need[leftChar] > 0 && window[leftChar] < need[leftChar] {
				formed--
			}

			left++
		}
	}

	if minLen == int(1e9) {
		return ""
	}

	return s[start : start+minLen]
}
